### ISSUE TITLE
FLOW reports using BIRT

### ISSUE THREAD

**Issue by stellanl** [akvo/akvo-product-design#130]

**Comment by jonase** [comment]
![flow_relational_data_model](https://cloud.githubusercontent.com/assets/101023/11278841/e0777d46-8ef5-11e5-8933-49f6b4c05aa0.png)

**Comment by jonase** [comment]
@mtwestra @iperdomo @stellanl The above diagram is not a final design but perhaps a starting point. 

Note that we currently don't have all the information necessary in the unified log to build a database with this exact design. In particular the `question_group` data is missing because at the time we implemented the data change capturing the responses were not as coupled to question groups as they are today due to repeatable **question groups**.

**Comment by jonase** [comment]
I just realized that we do have question_group data in the event log. The only thing that's missing is an `isRepeatable` property.

**Comment by stellanl** [comment]
That looks like a nice and clean structure.
Some questions:
Don't questions have an order within a question group?
Is this (except for names) a subset of the internal data structure of FLOW?
If not, then I suggest that the template generator could read from this database, making it just as independent of GAE internals as the BIRT engine is.

In either case, we need a test database with this design and some typical data in it (that will exercise the RQG machinery).

**Comment by jonase** [comment]
> Don't questions have an order within a question group?

yes, questions do have an `order` property. If we don't currently capture it in the unilog we should add it

> Is this (except for names) a subset of the internal data structure of FLOW?

This pretty closely aligns with the FLOW data model. Some bits and pieces could still be missing (for example question `order` which you already noticed).

> If not, then I suggest that the template generator could read from this database, making it just as independent of GAE internals as the BIRT engine is.

I agree. @mtwestra, @iperdomo what's your opinion on this. I know @mtwestra's proof of concept used the GAE datastore but I think it would be better to keep it out of GAE if possible.

> In either case, we need a test database with this design and some typical data in it (that will exercise the RQG machinery).

The database will be built by a unilog consumer. I'm not sure when I get the time to implement this. In the mean time we could perhaps manually create a test database with some test data?

**Comment by stellanl** [comment]
Yes, I should have phrased that better; I want to make a small DB manually with data that is a realistic challenge for the template and report generators.
Besides having a RQG and a geolocation question, I guess strings like question texts should not be trivially short, have more than a-z as letters etc.
Do you have a feel for what the max size (ignoring extreme outliers) of questions and answers is, or is there an instance that has data that has been notably "hard" that I could peek at?

**Comment by stellanl** [comment]
I have created a trivial first cut of such a db on psql.test under the name 'flowtestrep';
I had to change order to _order (SQL keyword) and the json data types to text (we only run PostgreSQL 9.1).

**Comment by stellanl** [comment]
A simple architecture if we go for the non-GAE option:
1. A webapp, secured by Keycloak, that generates report template instances from URL/form parameters and the FLOW reporting database(s).
2. A file store for the templates. It outputs a list of available reports and the links used to run them.
It could be a part of the above webapp or a separate one. FIles are stored in:
 a. a separate Postgres database (leverages established backup routines etc.)
 b. a local filesystem directory (simple, but needs to be backed up)
 c. S3 or other cloud storage. 
3. The BIRT viewer web-app, secured by Keycloak.

Future optional extras might be caching of the output from 3, Email delivery, etc.

A fundamental design decision is how to handle the separate FLOW instances.

**Comment by stellanl** [comment]
I think the proposed DB needs more metadata.
I. e. the time when a form_instance was collected feels like it could be an important part of selecting reported content.

**Comment by stellanl** [comment]
We will also need to define the format of the json in reponse.value.
Especially how it changes with question type.

**Comment by jonase** [comment]
Here's a partial list, I'll add more later

#### `OPTION`

Single option, with `code`:
```
[{"text": "foo", "code": "f"}]
```
Multiple options, without `code`:
```
[{"text": "foo"}, {"text": "bar"}]
```
Multiple options with `code` and `isOther` response:
```
[{"text": "foo", "code": "f"},
 {"text": "bar", "code": "b"},
 {"text": "abc", "code": "o", "isOther": true}]
```
#### `CASCADE`
```
[{"value": "foo", "code": "f"}, 
 {"value": "bar", "code": "b"}]
```
#### `GEOSHAPE`
As defined by [geojson](http://geojson.org/)

**Comment by stellanl** [comment]
So for option and cascade it is always an array, but for geoshape it is an object?

**Comment by stellanl** [comment]
To produce reports on enumerator activity etc. we will need to add the name of the submitter to the form instance. Device id would be nice here, too.

**Comment by stellanl** [comment]
Unless we need more attributes, I suggest we implement the geoshape response.value and data_point.location as GeoJSON geometry objects, like this: 
{"type": "Point", "coordinates": [17.9518, 59.2904]}

**Comment by jonase** [comment]
Currently (in the GAE datastore) we store geojson responses in this format:
```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type":"Polygon",
        "coordinates": [
          [[-0.11443745344877243,48.54595097934586],
           [1.8109390139579773,48.41552855346365],
           [1.2909465283155441,46.476951499425695],
           [-0.11443745344877243,48.54595097934586]]
        ]
      },
      "properties": {
        "pointCount":"3",
        "length":"615354.75",
        "area":"15565310565.70"
      }
    }
  ]
}
```

The consumer could strip out the outer layers. @ichinaski do we ever create geojson responses with multiple `Features` in the `FeatureCollection`?

**Comment by stellanl** [comment]
I suggest we simply pass the GeoJSON through unchanged. But the the location data of a data_point needs to be defined.

**Comment by jonase** [comment]
I think we have a few options: 

`JSON(B)` with the format:

```json
{"type": "Point", "coordinates": [17.9518, 59.2904]}
```

or, given that the type will always be `Point` (afaik, @mtwestra, @iperdomo  can you confirm?):

```json
[17.9518, 59.2904]
```

But if the type is always `Point` we can also use two columns (for example `latitude` and `longitude` of type `double precision`). I think I would prefer the "two columns" approach and only use `JSON` columns when necessary.

**Comment by jonase** [comment]
![flow_relational_data_model](https://cloud.githubusercontent.com/assets/101023/12015332/944843e4-ad43-11e5-8d59-96684cb885dd.png)

**Comment by jonase** [comment]
@stellanl Below is a sql script that can be run with the command

    $ createdb reporting
    $ psql -d reporting --set tenant="akvoflow_uat1" -f resources/tables.sql

The idea is to separate the tenants using schemas (instead of keeping them in separate databases). I could not figure out how to quote the schema name which is why I used the underscore version `akvoflow_uat1` instead of `akvoflow-uat1`. Also note that the foreign key indexes are created explicitly (instead of `... form_id bigint references form(id) ...`) which means that the order in which the tables are populated does not matter (we could create a question before the form exists) which should be fine as this is a "derived" database. (@iperdomo feedback welcome if this is a good idea or not)

```sql
create schema :tenant;

create table :tenant.survey (
  id bigint primary key,
  display_text text,
  description text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create table :tenant.form (
  id bigint primary key,
  survey_id bigint not null,
  display_text text,
  description text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on :tenant.form (survey_id);

create table :tenant.data_point (
  id bigint primary key,
  form_id bigint not null,
  identifier text,
  latitude double precision,
  longitude double precision,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on :tenant.data_point (form_id);

create table :tenant.form_instance (
  id bigint primary key,
  data_point_id bigint not null,
  submitter text,
  submitted_at timestamp,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on :tenant.form_instance (data_point_id);

create table :tenant.question_group (
  id bigint primary key,
  form_id bigint not null,
  "order" int,
  repeatable boolean default false,
  display_text text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (form_id, "order")
);

create index on :tenant.question_group (form_id);

create table :tenant.question (
  id bigint primary key,
  question_group_id bigint not null,
  "type" text,
  "order" int,
  display_text text,
  identifier text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (question_group_id, "order")
);

create index on :tenant.question (question_group_id);

create table :tenant.response (
  id bigint primary key,
  form_instance_id bigint not null,
  question_id bigint not null,
  "value" jsonb not null,
  iteration int default 0,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (form_instance_id, question_id, iteration)
);

create index on :tenant.response (form_instance_id);
create index on :tenant.response (question_id);

```

**Comment by jonase** [comment]
@stellanl While writing the consumer to support the above data model I discovered an error in the diagram (and the corresponding script). A data point is associated with a survey and a form instance is associated with a form:

![flow_relational_data_model](https://cloud.githubusercontent.com/assets/101023/12083537/cc8299de-b2ad-11e5-95e8-aaf12b9dfab7.png)

I will update the script above to support this data model.

**Comment by stellanl** [comment]
Well, that should simplify some of the queries...
Do we have to have columns called "order"? It's the only SQL real reserved word, and so you must always remember that, and needs special handling when you generate the queries programmatically. Could it be "ordering", "display_order" or "sort_order" instead?

**Comment by jonase** [comment]
@stellanl yeap, we can change that if we come up with a good name for it. Of the ones you propose I like `display_order` the most

**Comment by stellanl** [comment]
We should also define the starting point of the iteration and display_order fields. I have assumed 0 so far, and also assumed responses to non-repeatable question groups have an iteration=0.

**Comment by iperdomo** [comment]
> But if the type is always Point we can also use two columns (for example latitude and longitude of type double precision). I think I would prefer the "two columns" approach and only use JSON columns when necessary.

@stellanl @jonase 

Let's use 2 columns, `latitude`, `longitude` (or their equivalent short `lat` and `lon`).

Notice that we also capture elevation on the GEO questions. I guess we don't want to throw away that value.

**Comment by iperdomo** [comment]
> The idea is to separate the tenants using schemas (instead of keeping them in separate databases).

@jonase why do we want to use PG schemas for tenant separation instead of dbs ?

**Comment by iperdomo** [comment]
> Do we have to have columns called "order"?

We only need to quote the column name, right? @jonase has some example where `"value"` is quoted

**Comment by jonase** [comment]
> Notice that we also capture elevation on the GEO questions. I guess we don't want to throw away that value.
I can add `elevation` as well 

> @jonase why do we want to use PG schemas for tenant separation instead of dbs ?
So we can use connection pooling. Databases in postgres can't share connections. Having a pooling library manage the connection is easy and makes the consumer much more performant than opening (and closing) a connection on each event.

> @jonase has some example where "value" is quoted
I don't think `value` (or `type`) needs to be quoted. They are SQL key words (according to http://www.postgresql.org/docs/current/static/sql-keywords-appendix.html) but not reserved in postgres. I will try it and see

**Comment by iperdomo** [comment]
> So we can use connection pooling. Databases in postgres can't share connections. Having a pooling library manage the connection is easy and makes the consumer much more performant than opening (and closing) a connection on each event.

This a different approach, we should research how easy is to backup/restore just one schema in another server (e.g. if we want to redistribute the load across different clusters). What are the other operational implications of this decision.

**Comment by iperdomo** [comment]
e.g.

> When -n is specified, pg_dump makes no attempt to dump any other database objects that the selected schema(s) might depend upon. Therefore, there is no guarantee that the results of a specific-schema dump can be successfully restored by themselves into a clean database.

http://www.postgresql.org/docs/current/static/app-pgdump.html

**Comment by jonase** [comment]
> This a different approach, we should research how easy is to backup/restore just one schema in another server (e.g. if we want to redistribute the load across different clusters). What are the other operational implications of this decision.

Yeap. Also taking into account that these are derived views on the data.. so the same approach might not be necessary compared to (for example) the unilog.

**Comment by stellanl** [comment]
For a single-instance-at-a-time consumer like the report generator handling schemas other than public is rather trivial since JDBC 9.4; you can specify it as a parameter on the connection URL

**Comment by jonase** [comment]
@stellanl, @iperdomo  Below is a new sql script partitioning tenants into postgres databases instead of schemas:

    $ createdb akvoflow-uat1
    $ psql -d akvoflow-uat1 -f tables.sql

where `tables.sql` is
```sql
create table survey (
  id bigint primary key,
  display_text text,
  description text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create table form (
  id bigint primary key,
  survey_id bigint not null,
  display_text text,
  description text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on form (survey_id);

create table data_point (
  id bigint primary key,
  survey_id bigint not null,
  identifier text,
  latitude double precision,
  longitude double precision,
  elevation double precision,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on data_point (survey_id);

create table form_instance (
  id bigint primary key,
  form_id bigint not null,
  data_point_id bigint not null,
  submitter text,
  submitted_at timestamp,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp
);

create index on form_instance (form_id);
create index on form_instance (data_point_id);

create table question_group (
  id bigint primary key,
  form_id bigint not null,
  display_order int,
  repeatable boolean default false,
  display_text text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (form_id, display_order)
);

create index on question_group (form_id);

create table question (
  id bigint primary key,
  question_group_id bigint not null,
  type text,
  display_order int,
  display_text text,
  identifier text,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (question_group_id, display_order)
);

create index on question (question_group_id);

create table response (
  id bigint primary key,
  form_instance_id bigint not null,
  question_id bigint not null,
  value jsonb not null,
  iteration int default 0,
  created_at timestamp default current_timestamp,
  updated_at timestamp default current_timestamp,
  unique (form_instance_id, question_id, iteration)
);

create index on response (form_instance_id);
create index on response (question_id);
```

Other changes include:
* `"order"` -> `display_order`
* Drop quotes around `value` and `type`
* Add `elevation` to data_point table

**Comment by stellanl** [comment]
Looks good, my test database has been modified to match.

**Comment by janagombitova** [comment]
@stellanl @jonase @mtwestra @iperdomo @muloem  I have spoken with Satya today on the structure of our raw data reports and his experiences in analysing the data. All the notes are in this document. I will add in the input from Ethel as well in here once I receive it from her. 

https://docs.google.com/document/d/1jjiky1YNEujCtvRj83OeeuZ1YyJakCMkBW3Pbgvk_cs/edit

We should use this input when it comes to structuring our raw data reports on how we will handle the related questions. Can you please take a look and add your remarks?

**Comment by stellanl** [comment]
I can now generate reports where the RQGs are formatted as the |-separated lists, "arrays", Satya proposed. We need to define how the OPTION reponses are formatted, esp. when there are multiple selections.

**Comment by stellanl** [comment]
Report Designer is not happy about the Vietnamese text of form 30656057 when it ends up in a column header. Need to find out what is permitted there.

**Comment by stellanl** [comment]
Several of the users want to be able to split the multiple option responses into columns, one for each option. To do this, the non-other option values must be available when generating the template. In theory we could scan the database for all options in the answers, but that would miss any options that are not in any answers (at that moment). Much better if we could store the options as part of the questions.

**Comment by jonase** [comment]
For DASH integration we will also need the folder structure so I've updated the db model to add this. The folders can be ignored if they are not needed.

![flow_relational_data_model](https://cloud.githubusercontent.com/assets/101023/14274307/2c3e40cc-fb19-11e5-860b-d49619bdee69.png)

**Comment by stellanl** [comment]
The OPTION response format only works when the results go into one report cell.
Presence/absence of a specific option is very hard to access from a report SQL query since it can be anywhere or nowhere in the JSON array. Users want to be able to present multiple-option responses as yes/no in separate columns, with an optional "other" column.
This can be done from the current data, (although the report definition will have a limited shelf-life, as the set of "used" options could potentially change anytime a new response arrives).
I see two ways to do it:
1. The report generator maintains a table of processed options, updated when the report is defined.
2. The option response has an alternative/additional "content-adressed" format.
